homeassistant/components/whirlpool/sensor.py

```diff
@@ -0,0 +1,180 @@
+"""The Washer/Dryer Sensor for Whirlpool Appliances."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import logging
+
+from whirlpool.washerdryer import MachineState, WasherDryer
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import STATE_UNKNOWN
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+
+from . import WhirlpoolData
+from .const import (
+    CYCLE_FUNC,
+    DOMAIN,
+    ICON_D,
+    ICON_W,
+    MACHINE_STATE,
+    TANK_FILL,
+    WASHER_STATE,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def washer_state(washer: WasherDryer) -> str | None:
+    """Determine correct states for a washer."""
+
+    machine_state = washer.get_machine_state()
+    machine_cycle = None
+    for count, func in zip(range(len(CYCLE_FUNC)), CYCLE_FUNC):
+        if func(washer):
+            machine_cycle = WASHER_STATE.get(count)
+            break
+
+    if washer.get_attribute("Cavity_OpStatusDoorOpen") == "1":
+        machine_cycle = "Door open"
+        return machine_cycle
+
+    if machine_state == MachineState.RunningMainCycle and machine_cycle:
+        return machine_cycle
+
+    return MACHINE_STATE.get(machine_state, STATE_UNKNOWN)
+
+
+@dataclass
+class WhirlpoolSensorEntityDescriptionMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable
+
+
+@dataclass
+class WhirlpoolSensorEntityDescription(
+    SensorEntityDescription, WhirlpoolSensorEntityDescriptionMixin
+):
+    """Describes Whirlpool Washer sensor entity."""
+
+
+SENSORS: tuple[WhirlpoolSensorEntityDescription, ...] = (
+    WhirlpoolSensorEntityDescription(
+        key="state",
+        name="state",
+        entity_registry_enabled_default=True,
+        icon=ICON_W,
+        has_entity_name=True,
+        value_fn=washer_state,
+    ),
+    WhirlpoolSensorEntityDescription(
+        key="timeremaining",
+        name="time remaining",
+        device_class=SensorDeviceClass.DURATION,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement="s",
+        entity_registry_enabled_default=True,
+        icon=ICON_W,
+        has_entity_name=True,
+        value_fn=lambda WasherDryer: WasherDryer.get_attribute(
+            "Cavity_TimeStatusEstTimeRemaining"
+        ),
+    ),
+    WhirlpoolSensorEntityDescription(
```

Comment:
This seems like it may be a candidate for `entity_category` of `diagnostic` since its a lower level detail.

Reply:
It is the estimated end time for the wash/dry cycle.  My wife uses it routinely to see how much time is left - why would that be diagnostic?

Reply:
My comment was meant to be on "Dispense Level". Are you talking about dispense level or time remaining? I may misunderstand dispense level and the difference between the two sensers

Reply:
I was talking about time remaining (oops) -  the dispense level is the level of detergent in the auto detergent dispenser.   I don't think that's diagnostic either?

Reply:
Yes, that makes sense, thanks for that context I was missing. (I see in https://www.whirlpool.com/blog/washers-and-dryers/what-is-whirlpool-load-and-go.html it says you load in 20-40 loads worth of detergent then this tells you when its low.)

Do you think "Dispense level" is the right name? Just as one data point, I didn't understand what it meant. Could also consider "detergent level".